src/main/java/org/candlepin/service/exception/product/ProductUnknownRetrievalException.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2009 - 2023 Red Hat, Inc.
+ *
+ * This software is licensed to you under the GNU General Public License,
+ * version 2 (GPLv2). There is NO WARRANTY for this software, express or
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.service.exception.product;
+
+
+/**
+ * The ProductUnknownRetrievalException is used to reject a product retrieval for an unexpected
+ * reason.
+ */
+public class ProductUnknownRetrievalException extends ProductServiceException {
```

Comment:
I would agree with this. All of our service exceptions should derive from a common exception. Maybe a pair of common exceptions for both checked and unchecked variations: ServiceException (checked) and ServiceRuntimeException (unchecked).

Reply:
I did want to keep the "UnknownServiceException" at least be categorized as to the service that threw it. Since the  end goal is to stop passing messages through exceptions, then this differentiation is useful.

Reply:
A quick look at the adapter level changes, I drafted up an exception tree that looks something like this:

```
ServiceAdapterException
    CloudRegistrationServiceException
        CloudRegistrationAuthorizationException
        MalformedCloudRegistrationException

    EntitlementCertServiceException

    IdentityCertServiceException

    OwnerServiceException

    ProductServiceException

    SubscriptionServiceException
        ExpiredTagException
        ExhaustedTagException (tag already used)
        InvalidTagException

    UserServiceException
        InvalidCredentialsException
        UnacceptedTermsException
        UnauthorizedException
        UserDisabledException
```

This may not be a complete capture of everything (some of the calls were a bit difficult to follow), but it gives the outline of what the exception tree should look like. The remaining question is whether or not we consider these true error states or actionable and thus whether or not they should be checked or unchecked exceptions. Sadly with a tree like this, we're kinda forced into one or the other barring some shenanigans I'm not aware of.

I'm kinda leaning toward unchecked exceptions, since the only thing we really do with them is translate them to the appropriate error code, and not catching one should probably result in a 500 level as a signal that we need to catch it. In either case we're going to need the try/catch blocks around these calls, so the boilerplate is still going to be a thing in many cases, but I can be convinced either way yet.